### ISSUE TITLE
Lack of documentation

### ISSUE THREAD

**Issue by jclusso** [morning-brew/sanity-ruby#19]
I was expecting that when building our a PORO I'd be able to get something where I can work with it like I would an ActiveRecord object. To my surprise, the documentation is very vague on how this works or if this gem even provides that. I was expecting that I'd get instances of my object using a where query as an array of objects.

I saw it says "Extensibility in overriding the wrapper of your API response results" which should allow me at the least to do some sort of overriding, but I also see no documentation around this. Would it be possible to provide documentation on how this should be done?

**Comment by dvmonroe** [comment]
Thanks for the note @jclusso!

Indeed the documentation around this specific feature is subpar. If you're feeling up to you it, we'd welcome some support in updating it!

But to answer your question directly (with a couple assumptions):

----

I'll assume you're working with a PORO that looks something like this:

```ruby
class Person < Sanity::Resource
  attribute :_id, default: ""
  attribute :_type, default: "person"
  mutatable only: %i(create delete)
  queryable
end
```

I'll also assume you're using the [where query documented](https://github.com/morning-brew/sanity-ruby#querying) in the README like so:

```ruby
Person.where(_id: "123")
```

And you're probably being returned something that looks like this:

```ruby
{"ms"=>12,
 "query"=>"*[_id == '123']",
 "result"=>
  [{"_createdAt"=>"2021-09-21T02:44:36Z",
    "_id"=>"123",
    "_rev"=>"3456",
    "_type"=>"person",
    "_updatedAt"=>"2022-03-31T16:52:41Z",
    "name"=>"Batman"
   }]
 }
```

But you want an array of Person objects (similar to how ActiveRecord works)....

You can see [here](https://github.com/morning-brew/sanity-ruby/blob/main/lib/sanity/http/where.rb#L9) that our where class implements methods from our [query](https://github.com/morning-brew/sanity-ruby/blob/main/lib/sanity/http/query.rb) module. And in our query module [this](https://github.com/morning-brew/sanity-ruby/blob/main/lib/sanity/http/query.rb#L45) is where we determine how we return the results to you:

```ruby
http.request(request).then do |result|
  data = JSON.parse(result.body)

  block_given? ? yield(result_wrapper.call(data)) : result_wrapper.call(data)
end
```

`result_wrapper` is the key finding here.

If we work our way back we'll see that anything that implements Sanity::Http::Query accepts `result_wrapper` [on initialize](https://github.com/morning-brew/sanity-ruby/blob/main/lib/sanity/http/query.rb#L29) and that result wrapper has a default which is returning you the JSON results you see today.

We now know that 1) `result_wrapper` is an allowed arg when making a query and 2) result_wrapper must respond to call.

So first, we need to create this serializer/results class. Here's a crude example:

```ruby
class PersonSerializer
  class << self
    def call(...)
      new(...).call
    end
  end

  attr_reader :results

  def initialize(args)
    @results = args["result"]
  end

  def call
    results.map { |result| Person.new(_id: result["_id"], _type: result["_type"]) }
  end
end
```

Running this in our console should then output an array of person objects. Something like:

```ruby
Person.where(_id: "123", result_wrapper: PersonSerializer)
=> [#<Person:0x66980 _id: "123", _type: "person">]
```

WDYT? Does this help?

**Comment by jclusso** [comment]
This is very helpful. Thanks for all of this. I did end up finding `result_wrapper` after some searching. Some questions / improvement ideas.

- I know this is a breaking change, but I think `result_wrapper` should be renamed to `results_wrapper` since it doesn't wrap one result. 
- Is there a reason the default `result_wrapper` doesn't default to an enumerable of the PORO? Would you be against this change?
- I think there should be a configuration for the PORO to set the `result_wrapper` automatically.

Currently what we're doing is mixing this with ActiveModel because I wanted some of the nice features of it. I feel like this gem could do a lot more automagically, but I do understand the current simplicity of things currently. Some things we're doing are:

- Automatically setting a `result_wrapper` in the PORO and slicing the attributes of each result. 
- Automatically transforming the result keys into snake case to be Ruby friendly so things like `createdAt` are `created_at`.
- Automatically stripping the underscore from the beginning of system fields.

Let me know your thoughts.

**Comment by dvmonroe** [comment]
> I know this is a breaking change, but I think result_wrapper should be renamed to results_wrapper since it doesn't wrap one result.

We were following the nomenclature from Sanity. They return a singular `result` as the key to the results value. 

I'm not opposed to aliasing to support the singular or plural depending on the end user's desire. Aliasing would produce a simple, non-breaking change. 

> Is there a reason the default result_wrapper doesn't default to an enumerable of the PORO? Would you be against this change?

In the haste of building this gem it's just not something we prioritized. PRs welcome!

> I think there should be a configuration for the PORO to set the result_wrapper automatically.

100%. There are definitely non-breaking changes that we could make for this to be easier. One option that comes to mind is a macro that defines the result wrapper serializer.

-----

Hearing about your use case is neat! Great to hear others are finding this gem somewhat useful.

Your second and third points about transforming the sanity data to match ruby/AM expectations is interesting.

Perhaps the `attribute` macro could take an arg that specifies the "transformation". Something like:

```ruby
class Person < Sanity::Resource
  attribute :_id, tranform: :id, default: ""
end
```

**Comment by jclusso** [comment]
@dvmonroe I understand everything you're saying and like your ideas.

I think transform idea on the attribute you've got is nice. I'm a Mongo fan so I'd probably have done this since it's not abundantly clear that you can use `#id` on a person object.

```ruby
class Person < Sanity::Resource
  results_wrapper Person # Sanity::Resource would need to define everything for this to work.
  attribute :_id, alias: :id, default: ""
end
```

I'd also like to add `#attributes` since it's a nice method to be able to call for things.

**Comment by dvmonroe** [comment]
@swards do you mind elaborating a little bit and providing an short example of what you're trying to accomplish?

**Comment by jclusso** [comment]
I'm going to close this since I addressed this in #22.
